test/run.sh

```diff
@@ -1,16 +1,22 @@
 #!/bin/bash
+set -e # required to fail test suite when a single test fails
+set -x
 
 VAULT_VERSIONS=("1.0.0" "1.4.2")
 KV_BACKENDS=("KV1" "KV2")
 
-export DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+export DIR
+BATS_OS="$(command -v bats)" || true # allow to fail
```

Comment:
The intent of this is to try to use the system's default `bats` (if installed and in $PATH`) - am I correct with this assumption? 

I do not see a benefit doing this. What happens if the system has an incompatible version installed? Also, either way we must run `make get-bats` before running the test suits in order to download the support plugins which are being referenced - so we can also reference our downloaded bats-core to enhance stability. 
Am I forgetting sth important in my thoughts?

Reply:
This was an attempt to get it working on macOS for running the tests locally. Unfortunately, the binary downloaded with `make get-bats` does not work for Mac. I am fine skipping this change, since it currently brings no benefit.

Reply:
I do not have a Mac and cannot re-produce the issue. 
I could imagine the problem might be that bats is not installed via `install.sh`, and used purely from source. This works fine for Ubuntu, however there is an open issue for Mac https://github.com/bats-core/bats-core/issues/307
Is that the issue you are seeing? If so, then I can fix it quickly by using the `install.sh` to a local PREFIX. :crossed_fingers: 

Further, dockerizing bats would be nice indeed. Yet it would require a docker-in-docker approach, because the tests are managing the vault containers, which is why I was avoiding it.

Reply:
Installing `bats` on mac is not a problem - `brew install bats-core` works perfectly (but installs it into `/usr` - that is why I used the `command -v bats`). 

The problem I mentioned was with starting Vault in Docker - the `setup()` step was failing. I did not check the particular reason, but maybe I should. I find it a bit strange, because I could run the test suite in the previous (non-bats) version without problems.

Reply:
Ok got it. 
Out of curiosity: you mentioned an initial issue with the downloaded bats binary. When you are using the bats binary downloaded by `make get-bats`, does it show the issue mentioned above? 
I am asking because the whole point of this setup is to have an isolated bats env. No matter which version might be installed on the host system (e.g., via `brew install`), the test suite will work with the compatible version downloaded by `make get-bats`. If the issue above is what you are seeing, then I can try to fix it.

However, as you said this doesn't solve the docker issue. I am not really sure what the cause is here. If I am not mistaken, then gh actions for opensource also supports Mac builds for free. Maybe I should add that to the pipeline.

Reply:
What bats version is being installed by brew? Is it `1.2.0` or an older one?

Reply:
Yes, it is 1.2.0. Brew works quite well with regards to having recent versions.

Reply:
Okay, apparently when I install `bats` with `brew` and then use `make get-bats` then the version installed with `make get-bats` works - maybe some dependency library must have been missing?

Reply:
This are my errors when I try to run the tests with working `bats` installation (no matter which one):

```
$ VAULT_VERSION=1.0.0 test/bin/core/bin/bats test/special-tests

 ✗ vault-1.0.0 KV migrations with 'append'
   (from function `setup' in file test/special-tests/../util/util.bash, line 16,
    from function `bats_test_begin' in file test/bin/core/libexec/bats-core/bats-exec-test, line 97,
    in test file test/special-tests/kv-migrations-with-append.bats, line 5)
     `@test "vault-${VAULT_VERSION} KV migrations with 'append'" {' failed with status 125
 ✗ vault-1.0.0 KV migrations with 'cp'
   (from function `setup' in file test/special-tests/../util/util.bash, line 16,
    from function `bats_test_begin' in file test/bin/core/libexec/bats-core/bats-exec-test, line 97,
    in test file test/special-tests/kv-migrations-with-cp.bats, line 5)
     `@test "vault-${VAULT_VERSION} KV migrations with 'cp'" {' failed with status 125
 ✗ vault-1.0.0 KV migrations with 'mv'
   (from function `setup' in file test/special-tests/../util/util.bash, line 16,
    from function `bats_test_begin' in file test/bin/core/libexec/bats-core/bats-exec-test, line 97,
    in test file test/special-tests/kv-migrations-with-mv.bats, line 5)
     `@test "vault-${VAULT_VERSION} KV migrations with 'mv'" {' failed with status 125
 ✗ vault-1.0.0 whitespaces between parameters
   (from function `setup' in file test/special-tests/../util/util.bash, line 16,
    from function `bats_test_begin' in file test/bin/core/libexec/bats-core/bats-exec-test, line 97,
    in test file test/special-tests/params.bats, line 5)
     `@test "vault-${VAULT_VERSION} whitespaces between parameters" {' failed with status 125

4 tests, 4 failures
```

I will try to give it a deeper look and check what may be the cause

Reply:
Concerning the missing dependency: I am pretty sure this might be related to https://github.com/bats-core/bats-core/issues/307

Might be that `realpath` is not installed and gets installed with brew. Once the PR on bats-core is merged and released, I will update the version.

Reply:
I am somehow unable to revert it to check this, but this might have been the cause. I recall something about `real*` :)

Reply:
Okay, I have progress. After killing a running old Docker Vault image (4 days old, so it must have been 0.5.0 tests) I am spotting only a problem with hardcoded `linux` arch somewhere:

```
 ✗ vault-1.0.0 KV migrations with 'append'
   (from function `assert_success' in file test/special-tests/../bin/plugins/bats-assert/src/assert.bash, line 114,
    in test file test/special-tests/kv-migrations-with-append.bats, line 9)
     `assert_success' failed with status 126
   ==== case: append file from KV1 to KV2 ====

   -- command failed --
   status : 126
   output : $HOME/code/vsh/test/bin/core/libexec/bats-core/bats-exec-test: line 69: $HOME/code/vsh/test/util/../../build/vsh_linux_amd64: cannot execute binary file: Exec format error
   --
```

Reply:
This is now solved in 15bc81afbc2fd68a7cec9b932833bde6b25475cd